### ISSUE TITLE
CA1309 reported for string.Equals(string)

### ISSUE THREAD

**Issue by Bosch-Eli-Black** [dotnet/roslyn-analyzers#4627]
### Analyzer

**Diagnostic ID**: [CA1309](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1309): `Use ordinal StringComparison`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.1 (Latest)

### Describe the bug

This following piece of code produces a CA1309 warning:

```C#
string a = "abc";
string b = "def";

bool equals = a.Equals(b);
```

According to the [documentation for System.String.Equals()](https://docs.microsoft.com/en-us/dotnet/api/system.string.equals?view=net-5.0#System_String_Equals_System_String_), calling `Equals()` here "performs an ordinal (case-sensitive and culture-insensitive) comparison," so generating a CA1309 warning here seems like undesired behavior.

## Additional context

.csproj file:

```xml
<Project Sdk="Microsoft.NET.Sdk">

<PropertyGroup>
  <TargetFramework>netcoreapp3.1</TargetFramework>
  <AnalysisMode>AllEnabledByDefault</AnalysisMode>
	
<!-- AnalysisLevel is supposed to default to "latest", but there's a bug where if AnalysisLevel isn't set,
        the AnalysisMode setting doesn't take effect. Setting AnalysisLevel to "latest" works around this.
        See https://github.com/dotnet/roslyn-analyzers/issues/4530 -->
  <AnalysisLevel>latest</AnalysisLevel>
</PropertyGroup>
```

**Comment by Evangelink** [comment]
I think the idea of this rule is not only to catch incorrect configurations but also to require an explicit intent to be written. So I would think that this isn't really a false-positive.

@mavasani WDYT? We can go into a middle-ground option by introducing a user-option that would prevent the issue on known good cases.

**Comment by Bosch-Eli-Black** [comment]
With `NetAnalyzers` v6.0.0, the original example:

```c#
string a = "abc";
string b = "def";

bool equals = a.Equals(b);
```

now produces two warnings:

- [CA1309](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1309): Use ordinal StringComparison
- [CA1307](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1307): Specify StringComparison for clarity

IMHO, this is fine, and we can close this 🙂  Thoughts?

**Comment by Youssef1313** [comment]
@Bosch-Eli-Black It looks like both CA1307 and CA1309 have some overlap, which is something I personally don't like.

@Evangelink @mavasani Can these two be consolidated, or do they have different designs that they should remain?

**Comment by Evangelink** [comment]
They are a little different in the sense that `CA1309` would also report if the `StringComparison` was set to something other than `Ordinal` or `OrdinalIgnoreCase`. But I agree that having both reporting at the same time is not great... I would vote for keeping `CA1309` in this specific case (whether we go for suppression or merging the two classes together) because if we only report `CA1307` we may end up in a case where the user fixes the issue to then get `CA1309`.

**Comment by Bosch-Eli-Black** [comment]
The only thing strange about only reporting CA1309 in this case is that the message ("Use ordinal StringComparision") feels a bit strange, because [String.Equals() already uses ordinal comparison](https://docs.microsoft.com/en-us/dotnet/api/system.string.equals?view=net-6.0#System_String_Equals_System_String_).

Could/should CA1309 be renamed to "Specify ordinal StringComparison for clarity"?

**Comment by Bosch-Eli-Black** [comment]
Following up on this 🙂

**Comment by ninedan** [comment]
Agree, the warning should be removed for string.Equals(string, string), forcing it will make code larger and slower.

**Comment by abatishchev** [comment]
> Agree, the warning should be removed for string.Equals(string, string), forcing it will make code larger and slower.

@ninedan,
Can you clarify, or perhaps rephrase? Which option is slower and which is faster?
